File: samples/msgext-northwind-inventory-csharp/MessageExtensions/DiscountedSearchCommand.cs
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Schema.Teams;
using NorthwindInventory.NorthwindDB;
using NorthwindInventory.Handlers;

namespace NorthwindInventory.MessageExtensions
{
    public static class DiscountedSearchCommand
    {
        public const string CommandId = "discountSearch";

        public static async Task<MessagingExtensionResponse> HandleTeamsMessagingExtensionQueryAsync(ITurnContext<IInvokeActivity> turnContext, MessagingExtensionQuery query, IConfiguration configuration, CancellationToken cancellationToken)
        {
            var categoryName = Utils.CleanupParam(query.Parameters?.FirstOrDefault(p => p.Name == "categoryName")?.Value as string);
            var productService = new ProductService(configuration);
            var products = await productService.GetDiscountedProductsByCategory(categoryName);

            try
            {
                var attachments = new List<MessagingExtensionAttachment>();

                foreach (var product in products)
                {
                    var preview = new HeroCard
                    {
                        Title = product.ProductName,
                        Subtitle = $"Supplied by {product.SupplierName} of {product.SupplierCity}<br />{product.UnitsInStock} in stock",
                        Images = [new(product.ImageUrl)]
                    }.ToAttachment();

                    var resultCard = CardHandler.GetEditCard(product);

                    var attachment = new MessagingExtensionAttachment
                    {
                        ContentType = resultCard.ContentType,
                        Content = resultCard.Content,
                        Preview = preview
                    };

                    attachments.Add(attachment);
                }

                var messagingExtensionsAttachments = attachments
                    .Select(a => new MessagingExtensionAttachment
                    {
                        ContentType = a.ContentType,
                        Content = a.Content,
                        Preview = a.Preview
                    })
                    .ToList();

                return new MessagingExtensionResponse
                {
                    ComposeExtension = new MessagingExtensionResult
                    {
                        Type = "result",
                        AttachmentLayout = "list",
                        Attachments = messagingExtensionsAttachments
                    }
                };
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return null;
            }
        }
    }
}

File: samples/msgext-northwind-inventory-python/message_extensions/discounted_search_command.py
from botbuilder.core import TurnContext, CardFactory
from botbuilder.schema import HeroCard, CardImage, CardAction
from botbuilder.schema.teams import MessagingExtensionQuery, MessagingExtensionResponse,MessagingExtensionAttachment, MessagingExtensionResult
from northwindDB.products import get_discounted_products_by_category
from adaptive_cards import cards

COMMAND_ID = "discountSearch"
query_count = 0

async def handle_teams_messaging_extension_query(context: TurnContext, query: MessagingExtensionQuery) -> MessagingExtensionResponse:
    global query_count

    # Seek the parameter by name, don't assume it's in element 0 of the array
    category_name = cleanup_param(next((param.get("value") for param in query.parameters if param.get("name") == 'categoryName'), None))

    print(f'💰 Discount query #{query_count + 1}: Discounted products with categoryName={category_name}')
    query_count += 1

    products = await get_discounted_products_by_category(category_name)

    print(f'Found {len(products)} products in the Northwind database')

    attachments = []
    for obj in products:
        hero_card = HeroCard(
            title=obj["ProductName"], tap=CardAction(type="invoke", value=obj),
            text=f'Avg discount {obj["AverageDiscount"]}%<br />Supplied by {obj["SupplierName"]} of {obj["SupplierCity"]}',
            images=[CardImage(url=obj["ImageUrl"])]
        )

        attachment = MessagingExtensionAttachment(
            content_type=CardFactory.content_types.adaptive_card,
            content=cards.edit_card(obj),
            preview=CardFactory.hero_card(hero_card),
        )
        attachments.append(attachment)
    return MessagingExtensionResponse(
        compose_extension=MessagingExtensionResult(
            type="result", attachment_layout="list", attachments=attachments
        )
    )


def cleanup_param(value: str) -> str:
    if not value:
        return ''
    result = value.strip()
    result = result.split(',')[0]  
    result = result.replace('*', '')
    return result